src/frontend/src/components/LogIn/index.tsx

```diff
@@ -1,61 +1,173 @@
 import React from 'react';
 import { connect } from 'react-redux';
+import { Link } from 'react-router-dom';
 import { AppState } from '../../redux/store';
 import { logIn } from '../../redux/system/actions';
-import Logo from '../Logo';
 import { SystemState } from '../../redux/system/types';
+import apiService, { Request } from '../../service/api-service';
 import history from '../History';
-import { Link } from 'react-router-dom';
+import Logo from '../Logo';
 
 interface Props {
   system: SystemState;
   logIn: typeof logIn;
 }
-interface State {}
+
+interface State {
+  email: string;
+  password: string;
+  invalidCredentials: boolean;
+  invalidResponse: boolean;
+  invalidEmail: boolean;
+  missingInputs: boolean;
+}
 
 class LogIn extends React.Component<Props, State> {
+  public state: State = {
+    email: '',
+    password: '',
+    invalidCredentials: false,
+    invalidResponse: false,
+    invalidEmail: false,
+    missingInputs: false
+  };
+
   constructor(props: Props) {
     super(props);
     this.logInNow = this.logInNow.bind(this);
   }
 
-  public logInNow(event: React.SyntheticEvent) {
-    this.props.logIn();
+  public handleChange = (e: React.BaseSyntheticEvent) => {
+    // See https://github.com/DefinitelyTyped/DefinitelyTyped/issues/26635 for why we're
+    // using the "any" type.
+    this.setState<any>({
+      [e.target.id]: e.target.value
+    });
+  };
+
+  public logInNow(event: React.BaseSyntheticEvent) {
     event.preventDefault();
     event.stopPropagation();
-    history.push('/oeci');
+
+    if (this.validateForm()) {
+      const request: Request = {
+        url: '/api/auth_token',
+        data: {
+          email: this.state.email.toLowerCase().trim(),
```

Comment:
Everywhere `this.state.email` is used in this Component, the value used is not actually `this.state.email` but some modification of it. This suggests creating a function like

```
email() { 
  return this.state.email.trim().toLowerCase();
}
```

and using that instead, rather than doing the trim in so many places.

Reply:
This was another byproduct of the async nature of `this.setState` so is now refactored!

https://www.freecodecamp.org/news/get-pro-with-react-setstate-in-10-minutes-d38251d1c781/